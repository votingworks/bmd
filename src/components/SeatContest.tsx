import React from 'react'
import styled from 'styled-components'
import { Contest, InputEvent, UpdateVoteFunction, Vote } from '../config/types'

import Keyboard from 'react-simple-keyboard'
import 'react-simple-keyboard/build/css/index.css'

import Button from './Button'
import Modal from './Modal'
import Prose from './Prose'
import { Text } from './Typography'

const ContestSection = styled.small`
  font-weight: 600;
  text-transform: uppercase;
`
const FieldSet = styled.fieldset``
const Legend = styled.legend`
  margin-bottom: 1rem;
  @media (min-width: 480px) {
    margin-left: 4rem;
  }
`
const Choices = styled.div`
  display: grid;
  grid-auto-rows: minmax(auto, 1fr);
  grid-gap: 0.75rem;
`
const Choice = styled('label')<{ isSelected: boolean }>`
  position: relative;
  cursor: pointer;
  display: grid;
  grid-template-columns: 3fr 1fr;
  align-items: center;
  border-radius: 0.125rem;
  background: ${({ isSelected }) => (isSelected ? '#028099' : 'white')};
  color: ${({ isSelected }) => (isSelected ? 'white' : undefined)};
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0, 0, 0, 0.14),
    0 0.1875rem 0.0625rem -0.125rem rgba(0, 0, 0, 0.12),
    0 0.0625rem 0.3125rem 0 rgba(0, 0, 0, 0.2);
  transition: background 0.25s, color 0.25s;
  :focus-within {
    outline: -webkit-focus-ring-color auto 5px;
  }
  & > div {
    padding: 1rem;
  }
  & > input + div:before {
    content: '';
    position: absolute;
    left: 0;
    top: 0;
    bottom: 0;
    border-right: 1px solid lightgrey;
    width: 3rem;
    text-align: center;
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 2rem;
    border-radius: 0.125rem 0 0 0.125rem;
  }
  & > input:checked + div:before {
    content: 'âœ“';
    background: white;
    color: #028099;
    border-color: #028099;
  }
  & > div:first-of-type {
    padding-left: 4rem;
  }
  & .write-in-candidate-name {
    grid-column: span 2;
  }
`
const ChoiceInput = styled.input.attrs({
  type: 'checkbox',
})`
  margin-right: 0.5rem;
`

const WriteInCandidateForm = styled.div`
  background-color: lightgrey;
  margin: 1rem 0 -1rem;
  border-radius: 0.25rem;
  padding: 0.25rem;
`

const WriteInCandidateFieldSet = styled.fieldset`
  margin: 0.5rem 0.5rem 1rem;
`

const WriteInCandidateInput = styled.input.attrs({
  readOnly: true,
  type: 'text',
})`
  width: 100%;
  outline: none;
  box-shadow: 0 0 3px -1px rgba(0, 0, 0, 0.3);
  border: 1px solid darkgrey;
  padding: 0.25rem 0.35rem;
`

interface Props {
  contest: Contest
  vote: Vote
  updateVote: UpdateVoteFunction
}

interface State {
  attemptedVoteCandidateName: string
  writeInCandateModalIsOpen: boolean
  writeInCandidateName: string
  layoutName: string
}

const initialState = {
  attemptedVoteCandidateName: '',
  layoutName: 'default',
  writeInCandateModalIsOpen: false,
  writeInCandidateName: '',
}

class SeatContest extends React.Component<Props, State> {
  private keyboard: React.RefObject<Keyboard>
  constructor(props: Props) {
    super(props)
    let writeInCandidateName = ''
    if (typeof props.vote === 'string') {
      writeInCandidateName =
        !!props.vote &&
        props.contest.candidates
          .map(candidate => candidate.name)
          .includes(props.vote)
          ? ''
          : props.vote
    }
    this.state = {
      ...initialState,
      writeInCandidateName,
    }
    this.keyboard = React.createRef()
  }

  public selectCandidate = (selection: string) => {
    this.props.updateVote(this.props.contest.id, selection || '')
  }

  public handleUpdateSelection = (event: InputEvent) => {
    const target = event.target as HTMLInputElement
    const targetIsSelected = target.value === this.props.vote
    this.selectCandidate(targetIsSelected ? '' : target.value)
  }

  public handleChangeVoteAlert = (attemptedVoteCandidateName: string) => {
    this.setState({ attemptedVoteCandidateName })
  }

  public closeAttemptedVoteAlert = () => {
    this.setState({ attemptedVoteCandidateName: '' })
  }

  public handleSelectWriteInCandidate = () => {
    if (
      !!this.props.vote &&
      this.props.vote === this.state.writeInCandidateName
    ) {
      this.selectCandidate('')
    } else {
      this.toggleWriteInCandidateModal(true)
    }
  }
  public handleWriteInCandidateDisabledClick = () => {
    if (
      !!this.props.vote &&
      this.props.vote !== this.state.writeInCandidateName
    ) {
      this.handleChangeVoteAlert(
        this.state.writeInCandidateName || 'a write-in candidate'
      )
    }
  }
  public normalizeName = (name: string) =>
    name
      .trim()
      .replace(/\t+/g, ' ')
      .replace(/\s+/g, ' ')
  public selectWriteInCandidate = () => {
    const writeInCandidateName = this.normalizeName(
      this.state.writeInCandidateName
    )
    this.setState({ writeInCandidateName })
    this.selectCandidate(writeInCandidateName)
    this.toggleWriteInCandidateModal(false)
  }
  public closeWriteInCandidateModal = () => {
    const writeInCandidateName = this.normalizeName(
      this.state.writeInCandidateName
    )
    if (writeInCandidateName.length === 0) {
      this.setState({ writeInCandidateName })
    } else {
      this.selectCandidate(writeInCandidateName)
    }
    this.toggleWriteInCandidateModal(false)
  }
  public toggleWriteInCandidateModal = (writeInCandateModalIsOpen: boolean) => {
    this.setState({ writeInCandateModalIsOpen })
  }

  public setKeyboardInput = () => {
    this.keyboard.current!.setInput!(this.state.writeInCandidateName)
  }

  public onKeyboardInputChange = (writeInCandidateName: string) => {
    this.setState({ writeInCandidateName })
  }

  public onKeyboardKeyPress = (button: string) => {
    if (button === '{shift}' || button === '{lock}') {
      this.handleKeyboardShiftKey()
    }
  }

  public handleKeyboardShiftKey = () => {
    const layoutName = this.state.layoutName

    this.setState({
      layoutName: layoutName === 'default' ? 'shift' : 'default',
    })
  }

  public render() {
    const { contest, vote } = this.props
    const { attemptedVoteCandidateName } = this.state
    const writeInCandidateIsChecked =
      !!vote && vote === this.state.writeInCandidateName
    return (
      <React.Fragment>
        <FieldSet>
          <Legend>
            {contest.section && (
              <ContestSection>{contest.section}</ContestSection>
            )}
            <Prose>
              <h1>{contest.title}</h1>
              <p>
                <strong>Vote for 1.</strong> You have selected{' '}
                {!!vote ? `1` : `0`}.
              </p>
            </Prose>
          </Legend>
          <Choices>
            {contest.candidates.map((candidate, index) => {
              const isChecked = candidate.name === vote
              const handleDisabledClick = () => {
                if (vote && !isChecked) {
                  this.handleChangeVoteAlert(candidate.name)
                }
              }
              return (
                <Choice
                  key={candidate.name}
                  htmlFor={candidate.name}
                  isSelected={isChecked}
                  onClick={handleDisabledClick}
                >
                  <ChoiceInput
                    autoFocus={isChecked || (index === 0 && !vote)}
                    id={candidate.name}
                    name={contest.id}
                    value={candidate.name}
                    onChange={this.handleUpdateSelection}
                    checked={isChecked}
                    disabled={!!vote && !isChecked}
                    className="visually-hidden"
                  />
                  <div>
                    <strong>{candidate.name}</strong>
                  </div>
                  <div>{candidate.party}</div>
                </Choice>
              )
            })}
            <Choice
              htmlFor="writeInCandidate"
              isSelected={writeInCandidateIsChecked}
              onClick={this.handleWriteInCandidateDisabledClick}
            >
              <ChoiceInput
                autoFocus={writeInCandidateIsChecked}
                id="writeInCandidate"
                name={contest.id}
                value="writeInCandidate"
                onChange={this.handleSelectWriteInCandidate}
                checked={writeInCandidateIsChecked}
                disabled={!!vote && !writeInCandidateIsChecked}
                className="visually-hidden"
              />
              <div className="write-in-candidate-name">
                {!!this.state.writeInCandidateName ? (
                  <strong>{this.state.writeInCandidateName}</strong>
                ) : (
                  <em>add a write-in candidate</em>
                )}
              </div>
            </Choice>
          </Choices>
        </FieldSet>
        <Modal
          isOpen={!!attemptedVoteCandidateName}
          content={
            <Prose>
              <Text>
                To vote for {attemptedVoteCandidateName}, first uncheck the vote
                for {vote}.
              </Text>
            </Prose>
          }
          actions={
            <Button primary autoFocus onClick={this.closeAttemptedVoteAlert}>
              Okay
            </Button>
          }
        />
        <Modal
          isOpen={this.state.writeInCandateModalIsOpen}
          onAfterOpen={this.setKeyboardInput}
          content={
            <div>
              <Prose>
                <h2>Write-In Candidate</h2>
                <Text>
                  Use this screen to vote for a person who is{' '}
                  <strong>not</strong> on the ballot.
                </Text>
              </Prose>
              <WriteInCandidateForm>
                <WriteInCandidateFieldSet>
                  <legend>
                    <label htmlFor="WriteInCandidateName">
                      <Prose>
                        <Text bold small>
                          Write-In Candidate Name for {contest.title}
                        </Text>
                      </Prose>
                    </label>
                  </legend>
                  <WriteInCandidateInput
                    id="WriteInCandidateName"
                    value={this.state.writeInCandidateName}
                    placeholder="Use the keyboard to enter a candidate name"
                  />
                </WriteInCandidateFieldSet>
                <Keyboard
                  ref={this.keyboard}
                  layoutName={this.state.layoutName}
                  theme={'hg-theme-default vs-simple-keyboard'}
                  onChange={this.onKeyboardInputChange}
                  onKeyPress={this.onKeyboardKeyPress}
                  useButtonTag
                />
              </WriteInCandidateForm>
            </div>
          }
          actions={
            <>
              <Button
                primary={
                  this.normalizeName(this.state.writeInCandidateName).length > 0
                }
                autoFocus
                onClick={this.selectWriteInCandidate}
                disabled={
                  this.normalizeName(this.state.writeInCandidateName).length ===
                  0
                }
              >
                Accept
              </Button>
              <Button onClick={this.closeWriteInCandidateModal}>Close</Button>
            </>
          }
        />
      </React.Fragment>
    )
  }
}

export default SeatContest
