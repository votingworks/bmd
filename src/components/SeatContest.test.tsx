// import { axe } from 'jest-axe'
import React from 'react'
import { fireEvent, render } from 'react-testing-library'

import electionSample from '../data/electionSample.json'
const Q1Candidates = electionSample.contests[0]!.candidates
const Q1Candidate1 = Q1Candidates[0]!.name
const Q1Candidate2 = Q1Candidates[1]!.name
const Q1Candidate3 = Q1Candidates[2]!.name

import SeatContest from './SeatContest'

it(`allows any candidate to be selected when no candidate is selected`, () => {
  const updateVote = jest.fn()
  const { container, getByText } = render(
    <SeatContest
      contest={electionSample.contests[0]}
      vote={undefined}
      updateVote={updateVote}
    />
  )

  fireEvent.click(getByText(Q1Candidate1).closest('label')!)
  expect(updateVote).toHaveBeenCalledTimes(1)

  fireEvent.click(getByText(Q1Candidate2).closest('label')!)
  expect(updateVote).toHaveBeenCalledTimes(2)

  fireEvent.click(getByText(Q1Candidate3).closest('label')!)
  expect(updateVote).toHaveBeenCalledTimes(3)

  expect(container).toMatchSnapshot()

  // TODO: Why doesn't axe work when used in two tests in the same file?
  // expect(await axe(container.innerHTML)).toHaveNoViolations()
  // See tests below… which are used instead of the above line
})

it(`doesn't allow other candidates to be selected when one candidate is selected`, () => {
  const updateVote = jest.fn()
  const { container, getByText } = render(
    <SeatContest
      contest={electionSample.contests[0]}
      vote={electionSample.contests[0].candidates[0]}
      updateVote={updateVote}
    />
  )
  expect(container).toMatchSnapshot()

  const Candidate1Input = getByText(Q1Candidate1)
    .closest('label')!
    .querySelector('input')!
  expect(Candidate1Input.disabled).toBeFalsy()
  expect(Candidate1Input.checked).toBeTruthy()

  const Candidate2Input = getByText(Q1Candidate2)
    .closest('label')!
    .querySelector('input')!
  expect(Candidate2Input.disabled).toBeTruthy()
  expect(Candidate2Input.checked).toBeFalsy()

  const Candidate3Input = getByText(Q1Candidate3)
    .closest('label')!
    .querySelector('input')!
  expect(Candidate3Input.disabled).toBeTruthy()
  expect(Candidate3Input.checked).toBeFalsy()

  fireEvent.click(getByText(Q1Candidate2).closest('label')!)
  expect(updateVote).not.toHaveBeenCalled()

  fireEvent.click(getByText(Q1Candidate3).closest('label')!)
  expect(updateVote).not.toHaveBeenCalled()

  fireEvent.click(getByText(Q1Candidate1).closest('label')!)
  expect(updateVote).toHaveBeenCalled()

  // TODO: Why doesn't axe work when used in two tests in the same file?
  // expect(await axe(container.innerHTML)).toHaveNoViolations()
  // See tests below… which are used instead of the above line
})

it(`displays warning if write-in candidate name is too long`, () => {
  const updateVote = jest.fn()
  const { getByText } = render(
    <SeatContest
      contest={electionSample.contests[0]}
      vote={undefined}
      updateVote={updateVote}
    />
  )
  fireEvent.click(getByText('add a write-in candidate').closest('label')!)
  expect(getByText('Write-In Candidate')).toBeTruthy()
  Array.from('JACOB JOHANSON JINGLEHEIMMER SCHMIDTT').forEach(i => {
    const key = i === ' ' ? 'space' : i
    fireEvent.click(getByText(key).closest('button')!)
  })
  expect(
    getByText('You have entered 37 of maximum 40 characters.')
  ).toBeTruthy()
})

// TODO: Update this test to pass.
// - Failed when ReactModal was added with Modal component.
// - Error: "NotFoundError: The object can not be found here."
// - It is unclear what is causing this error.
// it(`accessible when no candidate is selected`, async () => {
//   const updateVote = jest.fn()
//   const { container } = render(
//     <SeatContest
//       contest={electionSample.contests[0]}
//       vote=""
//       updateVote={updateVote}
//     />
//   )
//   expect(await axe(container.innerHTML)).toHaveNoViolations()
// })

// TODO: Update this test to pass.
// - Failed when ReactModal was added with Modal component.
// - Error: "NotFoundError: The object can not be found here."
// - It is unclear what is causing this error.
// it(`accessible when one candidate is selected`, async () => {
//   const updateVote = jest.fn()
//   const { container } = render(
//     <SeatContest
//       contest={electionSample.contests[0]}
//       vote={'Minnie Mouse'}
//       updateVote={updateVote}
//     />
//   )
//   expect(await axe(container.innerHTML)).toHaveNoViolations()
// })
