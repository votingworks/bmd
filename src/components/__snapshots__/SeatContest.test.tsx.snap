// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows any candidate to be selected when no candidate is selected 1`] = `
.c1 {
  line-height: 1.3;
  max-width: 66ch;
}

.c1 h1 {
  margin: 2rem 0 1rem;
}

.c1 h2 {
  margin: 1.5rem 0 1rem;
}

.c1 h3,
.c1 p {
  margin: 1rem 0;
}

.c1 h1 + p,
.c1 h2 + p,
.c1 h3 + p {
  margin-top: -0.75rem;
}

.c1 > :first-child {
  margin-top: 0;
}

.c1 > :last-child {
  margin-bottom: 0;
}

.c0 {
  margin: 0 0 1rem 4rem;
}

.c2 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
}

.c3 {
  position: relative;
  cursor: pointer;
  display: grid;
  grid-template-columns: 3fr 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

.c3:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c3 > div {
  padding: 1rem;
}

.c3 > input + div:before {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  border-right: 1px solid lightgrey;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
}

.c3 > input:checked + div:before {
  content: '✓';
  background: white;
  color: #028099;
  border-color: #028099;
}

.c3 > div:first-of-type {
  padding-left: 4rem;
}

.c3 .write-in-candidate-name {
  grid-column: span 2;
}

.c4 {
  margin-right: 0.5rem;
}

<div>
  <fieldset
    class=""
  >
    <legend
      class="c0"
    >
      <div
        class="c1"
      >
        <h1>
          President
        </h1>
        <p>
          <strong>
            Vote for 1.
          </strong>
           You have selected
           
          0
          .
        </p>
      </div>
    </legend>
    <div
      class="c2"
    >
      <label
        class="c3"
        for="Minnie Mouse"
      >
        <input
          class="visually-hidden c4"
          id="Minnie Mouse"
          name="president"
          type="checkbox"
          value="Minnie Mouse"
        />
        <div>
          <strong>
            Minnie Mouse
          </strong>
        </div>
        <div>
          Mouse
        </div>
      </label>
      <label
        class="c3"
        for="Mickey Mouse"
      >
        <input
          class="visually-hidden c4"
          id="Mickey Mouse"
          name="president"
          type="checkbox"
          value="Mickey Mouse"
        />
        <div>
          <strong>
            Mickey Mouse
          </strong>
        </div>
        <div>
          Mouse
        </div>
      </label>
      <label
        class="c3"
        for="Donald Duck"
      >
        <input
          class="visually-hidden c4"
          id="Donald Duck"
          name="president"
          type="checkbox"
          value="Donald Duck"
        />
        <div>
          <strong>
            Donald Duck
          </strong>
        </div>
        <div>
          Duck
        </div>
      </label>
      <label
        class="c3"
        for="writeInCandidate"
      >
        <input
          class="visually-hidden c4"
          id="writeInCandidate"
          name="president"
          type="checkbox"
          value="writeInCandidate"
        />
        <div
          class="write-in-candidate-name"
        >
          <em>
            add a write-in candidate
          </em>
        </div>
      </label>
    </div>
  </fieldset>
</div>
`;

exports[`doesn't allow other candidates to be selected when one candidate is selected 1`] = `
.c1 {
  line-height: 1.3;
  max-width: 66ch;
}

.c1 h1 {
  margin: 2rem 0 1rem;
}

.c1 h2 {
  margin: 1.5rem 0 1rem;
}

.c1 h3,
.c1 p {
  margin: 1rem 0;
}

.c1 h1 + p,
.c1 h2 + p,
.c1 h3 + p {
  margin-top: -0.75rem;
}

.c1 > :first-child {
  margin-top: 0;
}

.c1 > :last-child {
  margin-bottom: 0;
}

.c0 {
  margin: 0 0 1rem 4rem;
}

.c2 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
}

.c5 {
  position: relative;
  cursor: pointer;
  display: grid;
  grid-template-columns: 3fr 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

.c5:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c5 > div {
  padding: 1rem;
}

.c5 > input + div:before {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  border-right: 1px solid lightgrey;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
}

.c5 > input:checked + div:before {
  content: '✓';
  background: white;
  color: #028099;
  border-color: #028099;
}

.c5 > div:first-of-type {
  padding-left: 4rem;
}

.c5 .write-in-candidate-name {
  grid-column: span 2;
}

.c3 {
  position: relative;
  cursor: pointer;
  display: grid;
  grid-template-columns: 3fr 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.125rem;
  background: #028099;
  color: white;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

.c3:focus-within {
  outline: -webkit-focus-ring-color auto 5px;
}

.c3 > div {
  padding: 1rem;
}

.c3 > input + div:before {
  content: '';
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  border-right: 1px solid lightgrey;
  width: 3rem;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 2rem;
  border-radius: 0.125rem 0 0 0.125rem;
}

.c3 > input:checked + div:before {
  content: '✓';
  background: white;
  color: #028099;
  border-color: #028099;
}

.c3 > div:first-of-type {
  padding-left: 4rem;
}

.c3 .write-in-candidate-name {
  grid-column: span 2;
}

.c4 {
  margin-right: 0.5rem;
}

<div>
  <fieldset
    class=""
  >
    <legend
      class="c0"
    >
      <div
        class="c1"
      >
        <h1>
          President
        </h1>
        <p>
          <strong>
            Vote for 1.
          </strong>
           You have selected
           
          1
          .
        </p>
      </div>
    </legend>
    <div
      class="c2"
    >
      <label
        class="c3"
        for="Minnie Mouse"
      >
        <input
          checked=""
          class="visually-hidden c4"
          id="Minnie Mouse"
          name="president"
          type="checkbox"
          value="Minnie Mouse"
        />
        <div>
          <strong>
            Minnie Mouse
          </strong>
        </div>
        <div>
          Mouse
        </div>
      </label>
      <label
        class="c5"
        for="Mickey Mouse"
      >
        <input
          class="visually-hidden c4"
          disabled=""
          id="Mickey Mouse"
          name="president"
          type="checkbox"
          value="Mickey Mouse"
        />
        <div>
          <strong>
            Mickey Mouse
          </strong>
        </div>
        <div>
          Mouse
        </div>
      </label>
      <label
        class="c5"
        for="Donald Duck"
      >
        <input
          class="visually-hidden c4"
          disabled=""
          id="Donald Duck"
          name="president"
          type="checkbox"
          value="Donald Duck"
        />
        <div>
          <strong>
            Donald Duck
          </strong>
        </div>
        <div>
          Duck
        </div>
      </label>
      <label
        class="c5"
        for="writeInCandidate"
      >
        <input
          class="visually-hidden c4"
          disabled=""
          id="writeInCandidate"
          name="president"
          type="checkbox"
          value="writeInCandidate"
        />
        <div
          class="write-in-candidate-name"
        >
          <em>
            add a write-in candidate
          </em>
        </div>
      </label>
    </div>
  </fieldset>
</div>
`;
